### ISSUE TITLE
Conflicts with airbnb preset

### ISSUE THREAD

**Issue by jtomaszewski** [davidje13/neutrino-typescript-eslint#1]
After scaffolding a neutrino react app with jest, airbnb, typescript and typescript-eslint preset, I get following errors after renaming example `index.js` and `App.js`. files to `index.tsx` `App.tsx`:

```
WARNING in ./src/index.tsx
Module Warning (from ./node_modules/eslint-loader/dist/cjs.js):
error: Missing file extension "tsx" for "./App" (import/extensions) at src/index.tsx:3:17:
  1 | import React from 'react';
  2 | import { render } from 'react-dom';
> 3 | import App from './App';
    |                 ^
  4 |
  5 | render(<App />, document.getElementById('root'));
  6 |


error: JSX not allowed in files with extension '.tsx' (react/jsx-filename-extension) at src/index.tsx:5:8:
  3 | import App from './App';
  4 |
> 5 | render(<App />, document.getElementById('root'));
    |        ^
  6 |


2 errors found.

WARNING in ./src/App.tsx
Module Warning (from ./node_modules/eslint-loader/dist/cjs.js):
warning: Missing return type on function (@typescript-eslint/explicit-function-return-type) at src/App.tsx:6:13:
  4 |
  5 | const message = 'Welcome to tf-portal-neutrino';
> 6 | const App = () => (
    |             ^
  7 |   <div className="App">
  8 |     <h1>{message}</h1>
  9 |   </div>


error: JSX not allowed in files with extension '.tsx' (react/jsx-filename-extension) at src/App.tsx:7:3:
   5 | const message = 'Welcome to tf-portal-neutrino';
   6 | const App = () => (
>  7 |   <div className="App">
     |   ^
   8 |     <h1>{message}</h1>
   9 |   </div>
  10 | );
```

What do you think we should do?

1. create a separate airbnb-typescript preset (just as the community has done with eslint configs, see https://www.npmjs.com/package/eslint-config-airbnb and https://www.npmjs.com/package/eslint-config-airbnb-typescript ), or

2. detect if airbnb preset is enabled, and if yes, tweak some of the eslint options

P.S. As for the solution no 1, I've found this: https://github.com/ConduitVC/neutrino-presets/blob/master/packages/airbnb/index.js .

**Comment by jtomaszewski** [comment]
For now as a quick workaround I have just set up the eslint config in an ordinary way and extended it ( https://github.com/infullmobile/twofront-eslint-config ) . Linking just in case somebody wants to reuse it.

**Comment by davidje13** [comment]
Yeah in my own projects I've been adding a [huge amount of eslint config](https://github.com/davidje13/collection-storage/blob/master/.neutrinorc.js#L15) on top of the airbnb preset (necessities mixed with per-project styles), but that's pretty nasty and a lot should be fixed here. Right now it doesn't even disable all the rules which don't work with typescript and have replacements (e.g. `'babel/semi': ['error']` should become `'@typescript-eslint/semi': ['error']`).

You list a few rules:

- `react/jsx-filename-extension`: Looks like this should be intercepted and have `tsx` added to the configured whitelist, similar to the code already in place for the import resolver.
- `import/extensions`: Airbnb explicitly configures [`js`, `jsx` and `mjs` differently](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/imports.js#L139), so again this would need to be intercepted (perhaps by default-copying the `js` config for `ts` and `jsx` for `tsx`?)
- `@typescript-eslint/explicit-function-return-type`: no conflict here and no equivalent JS rule, so I'd leave this to the individual project to decide if the rule should be disabled.

I think my preference would be what you proposed as \#2, since it wouldn't add much overhead to this package and would make it less confusing / more modular to use. In fact I'd prefer if any solution could avoid being airbnb-specific, so that it can "convert" any arbitrary base config. This could replace the use of `@typescript-eslint/eslint-recommended` (which just [disables some `eslint-recommended` rules without migrating their config](https://github.com/typescript-eslint/typescript-eslint/blob/d3470c963eb436d9e5128301d4579fb2b251de7c/packages/eslint-plugin/src/configs/eslint-recommended.ts); it doesn't even disable all the problematic rules, just the ones enabled by `@eslint/recommended`).

So I suggest the following changes:

- automatically add `tsx` to the `react/jsx-filename-extension` rule config if present, and similar for `import/extensions` (probably others will need this too; to-be-discovered)
- perform automatic conversion of everything listed in [this section of the docs](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#extension-rules) (including copying any configuration over), replacing the use of `@typescript-eslint/eslint-recommended`
- all auto-migration would only apply if there is no configured typescript-aware rule already (i.e. if the user says `'babel/semi': ['error'], '@typescript-eslint/semi': ['warn']`, the result would be `'babel/semi': ['off'], '@typescript-eslint/semi': ['warn']` not `'babel/semi': ['off'], '@typescript-eslint/semi': ['error']`
- all migrated rules apply only to ts/tsx files, not js/jsx (and any disabled rules are only disabled for ts/tsx too)

Would that be enough for your use case?

**Comment by relnod** [comment]
For people looking for a quick workaround, it is also possible to just disable the eslint rule, while still using the rest of the airbnb config.
```javascript
   ...
    airbnb({
      eslint: {
        baseConfig: {
          rules: {
            'react/jsx-filename-extension': 'off',
            'import/extensions': [
              'error',
              'ignorePackages',
              {
                'js': 'never',
                'jsx': 'never',
                'ts': 'never',
                'tsx': 'never'
              }
            ]

          }
        }
      }
    }),
   ...
```

**Comment by davidje13** [comment]
I have just released version 1.1.0 which adds automatic conversion of rules (see [adaptations.js](https://github.com/davidje13/neutrino-typescript-eslint/blob/master/adaptations.js) for specifics).

This is **very beta**. Please report any issues you see with the automatic rule conversion. I covered everything discussed in my last comment but there are probably mistakes and missing conversions.

**Comment by jtomaszewski** [comment]
Nice. Good work. I won't be able to confirm it at the moment as in our case we'll stay by our current config, but after a brief look at code changes, it looks good.

I will definitely check it out in my next project when such comes.

We might close this issue now I think. ;)
